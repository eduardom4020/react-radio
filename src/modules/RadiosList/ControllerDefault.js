import { useState, useCallback, useRef, useEffect } from 'react';

const ITEM_HEIGHT = 82;
const EXPANDED_ITEM_HEIGHT = 216;

const useController = props => {
    const { radios, web } = props;

    const [ radio, setRadio ] = useState(null);
    const [ position, setPosition ] = useState(null);
    const ref = useRef();
    
    // TEST: test if radios props is valid, if not reject it:
    if(typeof(radios) !== 'object' || ( typeof(radios) === 'object' && radios.constructor !== Object )) {
        return {
            dataValues: [],
            selectRadio: ()=>null,
            clearSelection: ()=>null,
            radio: {},
            ref
        };
    }

    const selectRadio = useCallback(
        (id) => {
            const radioItem = radios[id];

            if(radioItem) {
                setRadio(radioItem);
                setPosition(id);
            }
        }, 
        [ radios, setRadio ]
    );

    const clearSelection = useCallback(
        () => {
            setRadio(null);
            setPosition(null);
        }, 
        [ setRadio ]
    );

    useEffect(
        () => {
            if(ref && ref.current && position != null && ref.current.scrollToOffset) {
                const top = position * ITEM_HEIGHT;
                ref.current.scrollToOffset({ animated: true, offset: top > 0 ? top : 0 });
            }
        }, 
        [ ref, position ]
    );

    return {
        dataValues: Object.values(radios),
        selectRadio,
        clearSelection,
        radio,
        ref
    };
};

export default useController;