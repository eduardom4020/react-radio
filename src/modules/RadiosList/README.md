# RadiosList
>This module is a compoent in wich all the application run. It joins toolbar, radio itens and footer to make the entire application work. It has a 'global state' of the application and completes the click interaction by sending information to footer and the other itens.


In data folder we have a data source of the radios that is imported by this component and rendered to the user. This separation makes easy to create new radios, just by manipulating a json file.