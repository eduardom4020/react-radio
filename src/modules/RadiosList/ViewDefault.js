import React from 'react';
import styled from 'styled-components';
import { FlatList } from 'react-native';
import { RadioFooter, RadioItem, RadioToolbar } from '~/src/components';
import { MEDIUM_ORANGE, METALIC_BLACK } from '~/src/constants/colors';
import { APP_TITLE } from '~/src/constants/texts';
import useController from './ControllerDefault';
import DATA from '~/src/data/radios-data.json';

const StyledList = styled(FlatList)`
    padding: 24px 24px 32px 24px;
    width: 100%;
`;

const RadioListViewDefault = props => {
    const { radios: Radios, ...extraProps } = props;
    const radios = Radios || DATA;

    const { dataValues, radio, selectRadio, clearSelection, ref } = useController({...extraProps, radios});

    return (
        <>
            <RadioToolbar 
                color={MEDIUM_ORANGE}
                title={APP_TITLE}
                style={{paddingTop: 24}}
            />
            <StyledList
                ref={ref}
                data={dataValues}
                renderItem={({ item, index }) => (
                    <RadioItem 
                        {...item} 
                        selectedRadio={radio}
                        onToggle={(id, isActive) => {
                            if(radio && !isActive && radio.id === id) {
                                clearSelection();
                            } else {
                                selectRadio(id);
                            }
                        }}
                        borderVisible={Object.keys(DATA).length - 1 !== index}
                    />
                )}
                keyExtractor={item => item.id}
            />
            <RadioFooter 
                color={METALIC_BLACK}
                radioTitle={radio && radio.title}
            />
        </>
    );
};

export default RadioListViewDefault;