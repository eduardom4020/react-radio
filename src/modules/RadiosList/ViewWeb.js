import React from 'react';
import styled from 'styled-components';
import { FlatList } from 'react-native';
import { RadioFooter, RadioItem, RadioToolbar } from '~/src/components';
import { MEDIUM_ORANGE, METALIC_BLACK } from '~/src/constants/colors';
import { APP_TITLE } from '~/src/constants/texts';
import useController from './ControllerDefault';
import DATA from '~/src/data/radios-data.json';

const StyledList = styled.div`
    padding: 24px 24px 32px 6px;
    width: 100%;
    height: 55%;
    overflow-x: auto;
`;

const RadioListViewDefault = props => {
    const { radios: Radios, ...extraProps } = props;
    const radios = Radios || DATA;

    const { dataValues, radio, selectRadio, clearSelection, ref } = useController({...extraProps, radios});

    return (
        <>
            <RadioToolbar 
                color={MEDIUM_ORANGE}
                title={APP_TITLE}
            />
            <StyledList
                ref={ref}
            >
                {
                    dataValues.map((item, index, array) => (
                        <RadioItem 
                            id={item.id}
                            key={item.id}
                            {...item} 
                            selectedRadio={radio}
                            onToggle={(id, isActive) => {
                                if(radio && !isActive && radio.id === id) {
                                    clearSelection();
                                } else {
                                    selectRadio(id);
                                }
                            }}
                            borderVisible={array.length - 1 !== index}
                        />
                    ))
                }
            </StyledList>
            <RadioFooter 
                color={METALIC_BLACK}
                radioTitle={radio && radio.title}
            />
        </>
    );
};

export default RadioListViewDefault;