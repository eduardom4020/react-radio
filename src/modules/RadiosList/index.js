import View from './ViewDefault';

import WebView from './ViewWeb';
export const RadiosListWebView = WebView;

export default View;