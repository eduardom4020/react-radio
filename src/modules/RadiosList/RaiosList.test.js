import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import RadiosList from '~/src/modules/RadiosList';

describe('RadiosList', () => {
    test('Render?', () => {
        const component = shallow(
            <RadiosList />
        );

        expect(toJson(
            component, {
                noKey: false,
                mode: 'deep'
            }
        )).toMatchSnapshot();
    });
    
    test('Render with empty data provided?', () => {
        const component = shallow(
            <RadiosList radios={{}} />
        );

        expect(component.props().children[1].props.data).toMatchObject([]);
    });

    test('Render with different type of data?', () => {
        const componentBool = shallow(
            <RadiosList radios={true} />
        );

        expect(componentBool.props().children[1].props.data).toMatchObject([]);

        const componentString = shallow(
            <RadiosList radios={'some radio'} />
        );

        expect(componentString.props().children[1].props.data).toMatchObject([]);

        const componentArray = shallow(
            <RadiosList radios={[{id: 'radio1', title: 'Radio 1 FM', frequency: '1,0', tag: 'radioMaximumLogo'}]} />
        );

        expect(componentArray.props().children[1].props.data).toMatchObject([]);
    });

    test('Overrides default data with new one?', () => {
        const RADIOS = {
            radio1: {id: 'radio1', title: 'Radio 1 FM', frequency: '1,0', tag: 'radioMaximumLogo'},
            radio2: {id: 'radio2', title: 'Radio 2 FM', frequency: '2,0', tag: 'radioMaximumLogo'},
            radio3: {id: 'radio3', title: 'Radio 3 FM', frequency: '3,0', tag: 'radioMaximumLogo'}
        };

        const component = shallow(
            <RadiosList 
                radios={RADIOS}
            />
        );
        
        expect(component.props().children[1].props.data).toMatchObject(Object.values(RADIOS));
    });
});