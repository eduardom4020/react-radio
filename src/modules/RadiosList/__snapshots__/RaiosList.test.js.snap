// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadiosList Render? 1`] = `
<Fragment>
  <RadioToolbarViewDefault
    color="#eeae61"
    title="STATIONS"
  />
  <Styled(FlatList)
    data={
      Array [
        Object {
          "frequency": "66,6",
          "id": "1",
          "tag": "radioPutinLogo",
          "title": "Putin FM",
        },
        Object {
          "frequency": "101,2",
          "id": "2",
          "tag": "radioDribbbledLogo",
          "title": "Dribbble FM",
        },
        Object {
          "frequency": "99,4",
          "id": "3",
          "tag": "radioDogeLogo",
          "title": "Doge FM",
        },
        Object {
          "frequency": "87,1",
          "id": "4",
          "tag": "radioBalladsLogo",
          "title": "Ballads FM",
        },
        Object {
          "frequency": "142,2",
          "id": "5",
          "tag": "radioMaximumLogo",
          "title": "Maximum FM",
        },
      ]
    }
    keyExtractor={[Function]}
    renderItem={[Function]}
  />
  <RadioFooterViewDefault
    color="#191313"
    radioTitle={null}
  />
</Fragment>
`;
